lightning/src/util/base32.rs

```diff
@@ -0,0 +1,372 @@
+	-1, 18, -1, 25, 26, 27, 30, 29, 7, 31, -1, -1, -1, -1, -1, -1, -1, 24, 1, 12, 3, 8, 5, 6, 28,
+	21, 9, 10, -1, 11, 2, 16, 13, 14, 4, 22, 17, 19, -1, 20, 15, 0, 23,
+];
+
+/// Alphabet used for encoding and decoding.
+#[derive(Copy, Clone)]
+pub enum Alphabet {
+	/// RFC4648 encoding.
+	RFC4648 {
+		/// Whether to use padding.
+		padding: bool
+	},
+	/// Zbase32 encoding.
+	ZBase32
+}
+
+impl Alphabet {
+	/// Encode bytes into a base32 string.
+	pub fn encode(&self, data: &[u8]) -> Result<String, ()> {
+		let ret = match self {
+			Self::RFC4648 { padding } =>{
+				Self::rfc4648_encode(data, *padding)
+			},
+			Self::ZBase32 => {
+				Self::zbase32_encode(data)
+			}
+		};
+		match String::from_utf8(ret) {
+			Ok(s) => Ok(s),
+			Err(_) => Err(())
+		}
+	}
+
+	/// Decode a base32 string into a byte vector.
+	pub fn decode(&self, data: &str) -> Result<Vec<u8>, ()> {
+		match self {
+			Self::RFC4648 { .. } => {
+				Self::rfc4648_decode(data.to_ascii_uppercase().as_bytes())
+			},
+			Self::ZBase32 => {
+				Self::zbase32_decode(data.as_bytes())
+			},
+		}
+	}
+
+	fn rfc4648_encode(data: &[u8], padding: bool) -> Vec<u8> {
+		let cap = (data.len() + 3) / 4 * 5;
+		let mut ret = Self::encode_data(cap, data, RFC4648_ALPHABET);
+		if data.len() % 5 != 0 {
+			let len = ret.len();
+			let num_extra = 8 - (data.len() % 5 * 8 + 4) / 5;
+			if padding {
+				for i in 1..num_extra + 1 {
+					ret[len - i] = b'=';
+				}
+			} else {
+				ret.truncate(len - num_extra);
+			}
+		}
+		ret
+	}
+
+	fn zbase32_encode(data: &[u8])-> Vec<u8> {
+		let cap = (data.len() + 4) / 5 * 8;
+		let mut ret = Self::encode_data(cap, data, ZBASE_ALPHABET);
+		ret.truncate((data.len() * 8 + 4) / 5);
+		// Check that our capacity calculation doesn't under-shoot in fuzzing
+		#[cfg(fuzzing)]
+		assert_eq!(ret.capacity(), cap);
+		ret
+	}
+
+	fn validate_alphabet_char(&c: &u8, alphabet: [i8; 43], padding: Option<&mut usize>) -> Result<(), ()> {
+		if alphabet == RFC4648_INV_ALPHABET {
+			let padding = match padding {
+				Some(p) => p,
+				None => return Err(())
+			};
+			if c == b'=' {
+				*padding += 1;
+			} else if *padding > usize::MIN {
+				return Err(());
+			} else if !RFC4648_ALPHABET.contains(&c)  {
+				return Err(());
+			}
+			return Ok(());
+		}
+		if alphabet == ZBASE_INV_ALPHABET {
+			if !ZBASE_ALPHABET.contains(&c)  {
+				return Err(());
+			}
+			return Ok(());
+		}
+		return Err(());
+	}
+
+	fn rfc4648_decode(data: &[u8]) -> Result<Vec<u8>, ()> {
+		if data.len() == 0 {
+			return Ok(Vec::new())
+		}
+		if data.len() % 8 != 0 {
+			return Err(());
+		}
+		let mut unpadded_data_length = data.len();
+		data.iter().rev().take(6).for_each(|&c| {
+			if c == b'=' {
+				unpadded_data_length -= 1;
+			}
+		});
+		let output_length = unpadded_data_length * 5 / 8;
+		let cap = (output_length + 4) / 5 * 5;
+		let mut ret = Self::decode_data(cap, data, RFC4648_INV_ALPHABET)?;
+		ret.truncate(output_length);
+		#[cfg(fuzzing)]
+		assert_eq!(ret.capacity(), cap);
```

Comment:
its a bit challenging here because the `output_length` is defined based on different inputs because of the padding(for the `decode_data` we pass `data` but the `output_length` is defined by `unpadded_data`). also the zbase implementation is using the drain output to do some validation so that will require introducing some if to `decode_data`.